pkg/skaffold/kubernetes/portforward/kubectl_forwarder.go

```diff
@@ -20,105 +20,144 @@ import (
 	"bytes"
 	"context"
 	"fmt"
-	"net"
+	"io"
+	"os/exec"
 	"strings"
 	"time"
 
-	"github.com/pkg/errors"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+
 	"github.com/sirupsen/logrus"
-	"k8s.io/apimachinery/pkg/util/wait"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
-	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
 )
 
 type EntryForwarder interface {
-	Forward(parentCtx context.Context, pfe *portForwardEntry) error
+	Forward(parentCtx context.Context, pfe *portForwardEntry)
 	Terminate(p *portForwardEntry)
-	Monitor(*portForwardEntry, func())
 }
 
 type KubectlForwarder struct {
 	kubectl *kubectl.CLI
+	out     io.Writer
 }
 
-// Forward port-forwards a pod using kubectl port-forward
-// It returns an error only if the process fails or was terminated by a signal other than SIGTERM
-func (k *KubectlForwarder) Forward(parentCtx context.Context, pfe *portForwardEntry) error {
-	ctx, cancel := context.WithCancel(parentCtx)
-	// when retrying a portforwarding entry, it might already have a context running
-	if pfe.cancel != nil {
-		pfe.cancel()
-	}
-	pfe.cancel = cancel
-
-	cmd := k.kubectl.Command(ctx,
-		"port-forward",
-		"--pod-running-timeout", "5s",
-		fmt.Sprintf("%s/%s", pfe.resource.Type, pfe.resource.Name),
-		fmt.Sprintf("%d:%d", pfe.localPort, pfe.resource.Port),
-		"--namespace", pfe.resource.Namespace,
-	)
-	pfe.logBuffer = &bytes.Buffer{}
-	cmd.Stdout = pfe.logBuffer
-	cmd.Stderr = pfe.logBuffer
-
-	if err := cmd.Start(); err != nil {
-		if errors.Cause(err) == context.Canceled {
-			return nil
+// Forward port-forwards a pod using kubectl port-forward in the background
+// It kills the command on errors in the kubectl port-forward log
+// It restarts the command if it was not cancelled by skaffold
+// It retries in case the port is taken
+func (k *KubectlForwarder) Forward(parentCtx context.Context, pfe *portForwardEntry) {
+	go k.forward(parentCtx, pfe)
+}
+
+func (k *KubectlForwarder) forward(parentCtx context.Context, pfe *portForwardEntry) {
+	var notifiedUser bool
+	for {
```

Comment:
when reading this, it looks like this `forward` runs forever?

Reply:
it does, until the pfe gets cancelled (there are 3 return statements in the body, all around cancellation!)

Reply:
got lost in reading and saw only 1 return statement.
I still feel, all the 3 return statements are waiting for conditions that are
1. pfe.terminated. -  not sure when that happens
2. & 3. user enter ^C in `skaffold dev` .

Will it keep trying forever untill above two conditions are satisfied?

Reply:
well, if things are going well, it is going to wait in `cmd.Wait()` as long as kubectl portforward is running ...

Reply:
but yes, it does retry unless it is an explicit cancel from the skaffold process...

Reply:
Something like this: 

![image](https://user-images.githubusercontent.com/635073/62333350-04bbc300-b478-11e9-9a12-9f4c6d8f9d5b.png)

Reply:
btw this was generated on planttext.com :) 
``` 
@startuml

title Port Forward State Model
[*] --> CheckForTermination
CheckForTermination -down-> Cancelled
CheckForTermination -down-> TryCmdStart
TryCmdStart -down-> Cancelled
TryCmdStart --> CheckForTermination
TryCmdStart -down-> LogMonitoring

Cancelled -down-> [*]
state "kubectl port-forward logs are monitored" as LogMonitoring {
  TryCmdWait -down-> Cancelled
  TryCmdWait -up-> CheckForTermination
}

@enduml

```

Reply:
I wish I could put labels on the arrows - but the key thing is non-cancellation error scenarios go back to start, cancellation exists, no error continues to next step.